Documentation/PrettyPrinter.md

```diff
@@ -430,3 +430,192 @@ print it. This value is appended to the length array, and added to `total`.
+creates a new line from within the group, this value will no longer correspond
+to the correct global indentation. This is why we correct it with the relative
+indentation value of its containing group.
+
+### Break Tokens
+
+At `break` tokens we need to decide whether or not to create a line break. This
+occurs if the top of the `forceBreakStack` is true, or of the length of the
+`break` exceeds `spaceRemaining`. `lastBreakConsecutive` must be false.
+
+If we need to create a line break, we need to calcualte the indentation level
+and propagate that information to future `syntax` and `open` tokens. First read
+the top of `indentStack`, and add the indentation to the `break` token's offset
+value; this will be `lastBreakValue`. Set `lastBreak` and `lastBreakConsecutive`
+to true, and `lastBreakOffset` to the `break`'s offset. The `spaceRemaining` is
+set to the maximum line length minus `lastBreakValue`.
+
+If we do not need to create a new line, we print the `break`. We write out the
+number of spaces corresponding to the `break` token's size, and subtract the
+value of `size` from `spaceRemaining`. We then reset the `lastBreak` variables
+as we did for `syntax` tokens, except we leave `lastBreakConsecutive` unchanged.
+
+### Newline Tokens
+
+`newline` tokens do the same indentation calculations as `break` tokens do when
+they are creating a line break. Additionally, `newline` tokens also print the
+number of `\n` characters specified. We do not need to check lengths or
+`forceBreakStack`, since `newline` tokens always create line breaks.
+
+### Space Tokens
+
+`space` tokens behave similarly to `syntax` tokens. If the token is preceded by
+a `break` token that created a new line (`lastBreakConsecutive` is true), the
+number of spaces for the indent are printed, and the `lastBreak` variables are
+reset. Then, we print the number of spaces according to the `space` token's
+length, and adjust `spaceRemaining` accordingly.
+
+### Reset Tokens
+
+`reset` tokens reset the `lastBreak` variables. They do nothing else.
+
+### Comment Tokens
+
+`comment` tokens also be have similarly to `syntax` tokens. If
+`lastBreakConsecutive` is true, we apply the indent and reset the `lastBreak`
+variables as usual. Next, we print the contents of the `comment` tokens
+including any necessary delimiters. `spaceRemaining` is adjusted according to
+the length of the `comment`.
+
+### Verbatim Tokens
+
+When we encounter a `verbatim` token, we simply print it's contents and apply a
+global indentation according to `lastBreakValue`. We reset the `lastBreak`
+variables, and adjust `spaceRemaining` according to the token's length, which is
+equivalent to the maximum line width.
+
+## Differences from Oppen's Algorithm
+
+For those who might already be familiar with Oppen's pretty-printing algorithm,
+described below are ways in which SwiftFormat's pretty-printer differs from
+Oppen's.
+
+### Absence of a "stream"
```

Comment:
Random thought that just occurred to me: I wonder if we'd see meaningful performance improvements if we put `TokenStreamCreator` in a different thread and turned it into an _actual_ stream. Then it would run in parallel with the `print` thread, which would ask for tokens, queue them up in a ring buffer, and then output them as per the optimized algorithm.

Of course, the effort to implement that is fairly non-trivial and we have much bigger fish to fry right now. And it's also not clear whether our changes to the algorithm would make it even more difficult to implement that.

Reply:
That is certainly a possibility. I think we would only need to modify the *scan* phase of the algorithm, and *print* should remain the same. Fortunately for us, *print* is the more complicated of the two.